### PR TITLE
refactor: change `PoissonLikelihood` into a function with faster compilation, speed and low memory footprint

### PR DESCRIPTION
**Pull request by Qazalbash**
## Summary

We have changed `PoissonLikelihood` into `poisson_likelihood,` which compiles very fast and takes less memory than before.

## Related To

- #454 

## Description

In this PR, we have explored the use **padded arrays** for data to make their shapes same. We padded the data, reference priors, and an extra mask array to remember the padded values. When we have the same shape arrays, we could apply the `jax.vmap` transformation to calculate the likelihood of a single event. This was exhausting resources, therefore we moved to `jax.lax.scan` and it is working like a charm. We are getting faster compilation times, good speed, and low memory consumption.

## Additional Notes (Optional)

#525 contains partial changes and was merged accidentally.

## Summary by CodeRabbit

- **Refactor**
  - Replaced the class-based Poisson likelihood implementation with a function-based interface, affecting all related modules and usage patterns.
  - Updated all relevant scripts to use the new function-based Poisson likelihood, adjusting imports and variable usage accordingly.
  - Removed JAX device placement for reference priors and data arrays in affected modules.

- **Chores**
  - Updated pre-commit hook versions for improved tooling and checks.

- **Style**
  - Removed unused imports and added type ignore comments where necessary.

### CODE REVIEW

**Comment by Qazalbash** [comment]
@coderabbitai full review

**Comment by zeeshan5885** [comment]
### Stress Test Summary

Model: PowerLaw with Gaussian spin (no smoothing)
Dataset: 144 events with an error size of 12,000

Test Configurations and Observations:
	•	Initial Run:
	•	40 chains, 80 global steps, 80 local steps
	•	GPU Memory: 9.5 GB
	•	Runtime: 247 seconds
	•	**Test 2:**
	•	Increased global steps to 160
	•	Runtime remained approximately the same
	•	GPU Memory: 25 GB
	•	**Test 3:**
	•	Kept global steps at 160 (sufficient for convergence)
	•	Increased chains to 200
	•	GPU Memory: 34.5 GB
	•	**Test 4:**
	•	Increased chains to 400
	•	GPU Memory: 60 GB
	•	**Test 5:**
	•	Increased chains to 500
	•	GPU Memory: 70 GB
In general, with global steps or chains the memory demand is increasing linearly.
